traits/src/kem/slice.rs

```diff
@@ -0,0 +1,215 @@
+
+    /// Decapsulate a shared secret.
+    fn decaps(ss: &mut [u8], ct: &[u8], dk: &[u8]) -> Result<(), DecapsError>;
+}
+
+/// Error generating key with provided randomness
+#[derive(Debug)]
+pub enum KeyGenError {
+    /// Error generating key with provided randomness
+    InvalidRandomness,
+    /// The provided randomness has the wrong length
+    InvalidRandomnessLength,
+    /// The provided encapsulation key has the wrong length
+    InvalidEncapsKeyLength,
+    /// The provided decapulation key has the wrong length
+    InvalidDecapsKeyLength,
+    /// An unknown error occurred
+    Unknown,
+}
+
+/// Error indicating that encapsulating failed
+#[derive(Debug)]
+pub enum EncapsError {
+    /// Encapsulation key is invalid
+    InvalidEncapsKey,
+    /// Error encapsulating key with provided randomness
+    InvalidRandomness,
+    /// The provided randomness has the wrong length
+    InvalidRandomnessLength,
+    /// The provided encapsulation key has the wrong length
+    InvalidEncapsKeyLength,
+    /// The provided ciphertext has the wrong length
+    InvalidCiphertextLength,
+    /// The provided shared secret has the wrong length
+    InvalidSharedSecretLength,
+    /// An unknown error occurred
+    Unknown,
+}
+
+/// Error indicating that decapsulating failed
+#[derive(Debug)]
+pub enum DecapsError {
+    /// Ciphertext key is invalid
+    InvalidCiphertext,
+    /// Decapsulation key is invalid
+    InvalidDecapsKey,
+    /// The provided decapulation key has the wrong length
+    InvalidDecapsKeyLength,
+    /// The provided ciphertext has the wrong length
+    InvalidCiphertextLength,
+    /// The provided shared secret has the wrong length
+    InvalidSharedSecretLength,
+    /// An unknown error occurred
+    Unknown,
+}
+
+impl From<arrayref::KeyGenError> for KeyGenError {
+    fn from(value: arrayref::KeyGenError) -> Self {
+        match value {
+            arrayref::KeyGenError::InvalidRandomness => KeyGenError::InvalidRandomness,
+            arrayref::KeyGenError::Unknown => KeyGenError::Unknown,
+        }
+    }
+}
+
+impl From<arrayref::EncapsError> for EncapsError {
+    fn from(value: arrayref::EncapsError) -> Self {
+        match value {
+            arrayref::EncapsError::InvalidEncapsKey => EncapsError::InvalidEncapsKey,
+            arrayref::EncapsError::InvalidRandomness => EncapsError::InvalidRandomness,
+            arrayref::EncapsError::Unknown => EncapsError::Unknown,
+        }
+    }
+}
+
+impl From<arrayref::DecapsError> for DecapsError {
+    fn from(value: arrayref::DecapsError) -> Self {
+        match value {
+            arrayref::DecapsError::InvalidCiphertext => DecapsError::InvalidCiphertext,
+            arrayref::DecapsError::InvalidDecapsKey => DecapsError::InvalidDecapsKey,
+            arrayref::DecapsError::Unknown => DecapsError::Unknown,
+        }
+    }
+}
+
+#[macro_export]
+/// Implements [`Kem`] for any [`arrayref::Kem`]. Hides the impl in a module of the given name.
```

Comment:
```suggestion
/// Implements [`Kem`] for any [`arrayref::Kem`].
```
Or did I miss it somehow?